### ISSUE TITLE
ovs userspace build is failing on latest repo on windows platform

### ISSUE THREAD

**Issue by ankursh** [openvswitch/ovs-issues#26]
Steps for build:
./configure CC=./build-aux/cccl LD="`which link`" LIBS="-lws2_32" --prefix="C:/openvswitch/usr" --localstatedir="C:/openvswitch/var"   --sysconfdir="C:/openvswitch/etc"  --with-pthread="C:/pthread"

make
# 

Issue:
The build fails after following the steps.
It used to work fine till commit when we last tested it when branch was synced till commit:
222820c67cea17ed76024364dcdd08778d1232fc

some thing between following commits has broken the build
222820c67cea17ed76024364dcdd08778d1232fc - 6ebad850175a9891b27562ec4d8958aa478fdcd9

**Comment by nithinrajub** [comment]
Ankur, I am able to build successfully if I sync to the latest as well as to 6ebad850175a9891b27562ec4d8958aa478fdcd9.

I am using the following commands:
$ ./boot.sh
$ ./configure CC=./build-aux/cccl LD="`which link`" LIBS="-lws2_32" --prefix="C
:/openvswitch/usr" --localstatedir="C:/openvswitch/var" --sysconfdir="C:/openvs
witch/etc" --with-pthread="C:/pthread"
$ make

Did you build the Linux binaries in the same tree? That can be one source of the error.

You can try, make distclean and try building again.

**Comment by ssp0** [comment]
what is the error like?

**Comment by aserdean** [comment]
Did a git clone just now. Had a successful full build. Can you please paste the error?

**Comment by ankursh** [comment]
Let me try again after distclean. Could be a mistake on my side too.

The error was:
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') < ./rhel/openvswitch-fedora.spec
.in > openvswitch-fedora.spec.tmp || exit 1; if cmp -s openvswitch-fedora.spec.t
mp rhel/openvswitch-fedora.spec; then touch rhel/openvswitch-fedora.spec; rm ope
nvswitch-fedora.spec.tmp; else mv openvswitch-fedora.spec.tmp rhel/openvswitch-f
edora.spec; fi
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') \
                < ./xenserver/openvswitch-xen.spec.in > openvswitch-xen.spec.tmp
 || exit 1; \
        if cmp -s openvswitch-xen.spec.tmp xenserver/openvswitch-xen.spec; then
touch xenserver/openvswitch-xen.spec; rm openvswitch-xen.spec.tmp; else mv openv
switch-xen.spec.tmp xenserver/openvswitch-xen.spec; fi
-uThe system cannot find the file specified.
make[2]: **\* [distfiles] Error 1
make[2]: Leaving directory `/c/ovs_package/ovs_hyperv_code/code1'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory`/c/ovs_package/ovs_hyperv_code/code1'
make: **\* [all] Error 2
bash-3.1$

**Comment by aserdean** [comment]
I get the following:
        if cmp -s openvswitch-xen.spec.tmp xenserver/openvswitch-xen.spec; then
touch xenserver/openvswitch-xen.spec; rm openvswitch-xen.spec.tmp; else mv openv
switch-xen.spec.tmp xenserver/openvswitch-xen.spec; fi
make[3]: Entering directory `/c/temp/temp/ovs/datapath'
make[3]: Leaving directory`/c/temp/temp/ovs/datapath'
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/nicira-ext.h
touch include/openflow/nicira-ext.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-1.0.h
touch include/openflow/openflow-1.0.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-1.1.h
touch include/openflow/openflow-1.1.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-1.2.h
touch include/openflow/openflow-1.2.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-1.3.h
touch include/openflow/openflow-1.3.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-1.4.h
touch include/openflow/openflow-1.4.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow-common.h
touch include/openflow/openflow-common.hstamp
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27/python
 ./build-aux/check-structs -I./include include/openflow/openflow.h
touch include/openflow/openflow.hstamp
make[2]: Leaving directory `/c/temp/temp/ovs'
make[1]: Leaving directory`/c/temp/temp/ovs'

I see your missing a file but can not make out which one exactly.

**Comment by ssp0** [comment]
I may be totally off, but, are you compiling under the developer prompt?

> I see your missing a file but can not make out which one exactly.

It is hard to spot this, but I think the cause of the error is "-uThe system cannot find the file specified.". I have seen this error before, trying to remember what the problem was.

**Comment by ankursh** [comment]
I built it on 2 different machines and faced the same problem on both. 
Most likely its my build environment issue (since this is the first time i am using make, previously i used to make selected binaries as was mentioned in install.WINDOWS ).

Since its working fine for both alin and nithin, hence i am reducing the priority and will work on it.

**Comment by ssp0** [comment]
One other thing I can think of is, are you using python 2.7? Python 3.0 is not supported.

**Comment by aserdean** [comment]
@ankursh can you please put in the full log somewhere on pastebin, might be more helpful to pinpoint the actual problem.

**Comment by ankursh** [comment]
saurabh: Yes i am using python2.7
Alin: Sure i'll paste the full log on pastebin.

**Comment by aserdean** [comment]
Ankur,
Did you manage to get a full build on windows in the end?

Please close the issue or post the log if there still are problems.

**Comment by ankursh** [comment]
Hi Alin,
No on 2 of my machines build did not work (looks like it is an issue with my build environment, since it works for you and nithin). So i reduced the priority and assigned it to myself. Sorry i did not get time to repro and paste the logs on pastebin.
The reason i kept it open is so that if i have to update something in INSTALL.WINDOWS then i can use this issue.

But if you think its better to close it then we can do that as well.

**Comment by aserdean** [comment]
Sure no problem keep it opened until we find out what happened on your environment.

Again if you need help just post the log somewhere maybe we can identify what the problem is.

**Comment by ankursh** [comment]
Sure, thanks a lot for following up on this alin.

**Comment by mantri** [comment]
Hi,

I see this same error on the latest master. I am using the Developer Command Prompt for VS2013. I am using python 2.7 and pthreads library from ftp://sourceware.org/pub/pthreads-win32/pthreads-w32-2-9-1-release.zip (copied from appveyor.yml)

bash-3.1$ which link
/c/Program Files (x86)/Microsoft Visual Studio 12.0/VC/BIN/link.exe
bash-3.1$ which cl
/c/Program Files (x86)/Microsoft Visual Studio 12.0/VC/BIN/cl.exe
bash-3.1$

I used the following commands:
1. ./boot.sh
2. ./configure --with-debug CC=build-aux/cccl LD="`which link`" LIBS=-lws2_32 --prefix="C:/openvswitch/usr" localstatedir="C:/openvswitch/var"  --sysconfdir="C:/openvswitch/etc" --with-pthread=C:/pthread/Pre-built.2 --enable-ssl --with-openssl=C:/OpenSSL-Win32 --with-vstudioddk="Win8.1 Debug"

I also tried without the SSL flag but I get the same error seen by @ankursh 

Here is the sinppet of the make failure: 

# 

mv vtep/vtep-ctl.8.tmp vtep/vtep-ctl.8
PYTHONPATH=./python";"$PYTHONPATH PYTHONDONTWRITEBYTECODE=yes /c/Python27//pytho
n ./ovsdb/ovsdb-doc \
                 \
                --version=2.3.90 \
                ./vtep/vtep.ovsschema \
                ./vtep/vtep.xml > vtep/vtep.5.tmp && \
        mv vtep/vtep.5.tmp vtep/vtep.5
printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
t -_-\n' > version.py.tmp && \
        echo 'VERSION = "2.3.90"' >> version.py.tmp && \
        if cmp -s version.py.tmp python/ovs/version.py; then touch python/ovs/ve
rsion.py; rm version.py.tmp; else mv version.py.tmp python/ovs/version.py; fi
/usr/bin//perl ./build-aux/soexpand.pl -I. < utilities/ovs-lib.in | \
            sed \
                -e 's,[@]PKIDIR[@],C:/openvswitch/var/lib/openvswitch/pki,g' \
                -e 's,[@]LOGDIR[@],C:/openvswitch/var/log/openvswitch,g' \
                -e 's,[@]DBDIR[@],C:/openvswitch/etc/openvswitch,g' \
                -e 's,[@]PERL[@],/usr/bin//perl,g' \
                -e 's,[@]PYTHON[@],/c/Python27//python,g' \
                -e 's,[@]RUNDIR[@],C:/openvswitch/var/run/openvswitch,g' \
                -e 's,[@]VERSION[@],2.3.90,g' \
                -e 's,[@]localstatedir[@],C:/openvswitch/var,g' \
                -e 's,[@]pkgdatadir[@],C:/openvswitch/usr/share/openvswitch,g' \

```
            -e 's,[@]sysconfdir[@],C:/openvswitch/etc,g' \
            -e 's,[@]bindir[@],C:/openvswitch/usr/bin,g' \
            -e 's,[@]sbindir[@],C:/openvswitch/usr/sbin,g' \
            -e 's,[@]abs_top_srcdir[@],/c/ovs,g' \
        > utilities/ovs-lib.tmp
```

mv utilities/ovs-lib.tmp utilities/ovs-lib
\
        { sed -n -e '/%AUTHORS%/q' -e p < ./debian/copyright.in;   \
          sed '1,/^$/d' ./AUTHORS |                                \
                sed -n -e '/^$/q' -e 's/^/  /p';                           \
          sed -e '1,/%AUTHORS%/d' ./debian/copyright.in;           \
        } > debian/copyright
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') < ./rhel/openvswitch-kmod-rhel6.
spec.in > openvswitch-kmod-rhel6.spec.tmp || exit 1; if cmp -s openvswitch-kmod-
rhel6.spec.tmp rhel/openvswitch-kmod-rhel6.spec; then touch rhel/openvswitch-kmo
d-rhel6.spec; rm openvswitch-kmod-rhel6.spec.tmp; else mv openvswitch-kmod-rhel6
.spec.tmp rhel/openvswitch-kmod-rhel6.spec; fi
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') < ./rhel/openvswitch-kmod-fedora
.spec.in > openvswitch-kmod-fedora.spec.tmp || exit 1; if cmp -s openvswitch-kmo
d-fedora.spec.tmp rhel/openvswitch-kmod-fedora.spec; then touch rhel/openvswitch
-kmod-fedora.spec; rm openvswitch-kmod-fedora.spec.tmp; else mv openvswitch-kmod
-fedora.spec.tmp rhel/openvswitch-kmod-fedora.spec; fi
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') < ./rhel/openvswitch.spec.in > o
penvswitch.spec.tmp || exit 1; if cmp -s openvswitch.spec.tmp rhel/openvswitch.s
pec; then touch rhel/openvswitch.spec; rm openvswitch.spec.tmp; else mv openvswi
tch.spec.tmp rhel/openvswitch.spec; fi
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') < ./rhel/openvswitch-fedora.spec
.in > openvswitch-fedora.spec.tmp || exit 1; if cmp -s openvswitch-fedora.spec.t
mp rhel/openvswitch-fedora.spec; then touch rhel/openvswitch-fedora.spec; rm ope
nvswitch-fedora.spec.tmp; else mv openvswitch-fedora.spec.tmp rhel/openvswitch-f
edora.spec; fi
(printf '\043 Generated automatically -- do not modify!    -_\- buffer-read-only:
 t -_-\n' && sed -e 's,[@]VERSION[@],2.3.90,g') \
                < ./xenserver/openvswitch-xen.spec.in > openvswitch-xen.spec.tmp
 || exit 1; \
        if cmp -s openvswitch-xen.spec.tmp xenserver/openvswitch-xen.spec; then
touch xenserver/openvswitch-xen.spec; rm openvswitch-xen.spec.tmp; else mv openv
switch-xen.spec.tmp xenserver/openvswitch-xen.spec; fi
-uThe system cannot find the file specified.
make[2]: **\* [distfiles] Error 1
make[2]: Leaving directory `/c/ovs'
make[1]: *** [all-recursive] Error 1
make[1]: Leaving directory`/c/ovs'
make: **\* [all] Error 2

# bash-3.1$

**Comment by blp** [comment]
On Tue, Mar 17, 2015 at 02:25:00PM -0700, Hemanth Kumar Mantri wrote:

> -uThe system cannot find the file specified.
> make[2]: **\* [distfiles] Error 1

The "distfiles" target is failing, probably related to "sort -u".
Huh.  Try removing the @ from the second line here:

distfiles: Makefile
    @srcdirstrip=`echo "$(srcdir)" | sed 's/[].[^$$\\*]/\\\\&/g'`; \

and then rerunning the make, to get more information.

**Comment by shettyg** [comment]
I think a different 'sort' is in your path.

**Comment by mantri** [comment]
I seem to have the Windows sort.exe in my $PATH

bash-3.1$ which sort
/c/Windows/system32/sort.exe

bash-3.1$ diff /usr/bin/sort.exe /c/Windows/System32/sort.exe
Files /usr/bin/sort.exe and /c/Windows/System32/sort.exe differ

Here is my $PATH:

/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/CommonExtensions
/Microsoft/TestWindow:/c/Program Files (x86)/Microsoft SDKs/F#/3.1/Framework/v4.
0/:/c/Program Files (x86)/MSBuild/12.0/bin:/c/Program Files (x86)/Microsoft Visu
al Studio 12.0/Common7/IDE/:/c/Program Files (x86)/Microsoft Visual Studio 12.0/
VC/BIN:/c/Program Files (x86)/Microsoft Visual Studio 12.0/Common7/Tools:/c/Wind
ows/Microsoft.NET/Framework/v4.0.30319:/c/Program Files (x86)/Microsoft Visual S
tudio 12.0/VC/VCPackages:/c/Program Files (x86)/HTML Help Workshop:/c/Program Fi
les (x86)/Microsoft Visual Studio 12.0/Team Tools/Performance Tools:/c/Program F
iles (x86)/Windows Kits/8.1/bin/x86:/c/Program Files (x86)/Microsoft SDKs/Window
s/v8.1A/bin/NETFX 4.5.1 Tools/:/c/Windows/system32:/c/Windows:/c/Windows/System3
2/Wbem:/c/Windows/System32/WindowsPowerShell/v1.0/:/c/Program Files (x86)/Iomete
r.org/Iometer 2006.07.27:/c/Program Files (x86)/Windows Kits/8.1/Windows Perform
ance Toolkit/:/c/Program Files/Microsoft SQL Server/110/Tools/Binn/:/mingw/bin/:
/usr/bin/:/c/Python27/:/c/Program Files (x86)/Git/cmd

Is this not expected? Thanks in advance.

**Comment by EitanEliahu** [comment]
Yes, Windows has Sort,exe  too.
You want to try to have the mingw\bin  first in the PATH.
Eitan

**Comment by shettyg** [comment]
This is my windows system $PATH

%SystemRoot%\system32;%SystemRoot%;%SystemRoot%\System32\Wbem;%SYSTEMROOT%\System32\WindowsPowerShell\v1.0\;C:\Program Files (x86)\Windows Kits\8.1\Windows Performance Toolkit\;C:\Program Files\Microsoft SQL Server\110\Tools\Binn\;C:\MinGW\bin;C:\Mingw\msys\1.0\bin;C:\Program Files (x86)\Git\cmd;C:\pthread\dll\x86;C:\Python27;C:\OpenSSL-Win32

**Comment by mantri** [comment]
Thanks, a lot. The build succeeded now. I did this so that the /usr/bin comes first in the PATH.

bash-3.1$ export PATH=/usr/bin:$PATH
bash-3.1$ which sort
/usr/bin/sort.exe
bash-3.1$

This may be worth documenting in INSTALL.WINDOWS.md 

Thanks All.

**Comment by shettyg** [comment]
I am trying to understand what is it that you did that caused this situation. INSTALL.Windows.md asks you to do the following:
Add "C:\MinGW\bin" and "C:\Mingw\msys\1.0\bin" to PATH environment variable
of Windows.

Did you not do the above? Did you do it in a different way?

**Comment by EitanEliahu** [comment]
I think that mingw was initially in the PATH be default or the VS++ path was added after opening the mingw shell.

**Comment by blp** [comment]
On Tue, Mar 17, 2015 at 03:44:52PM -0700, Gurucharan Shetty wrote:

> I am trying to understand what is it that you did that caused this situation. INSTALL.Windows.md asks you to do the following:
> Add "C:\MinGW\bin" and "C:\Mingw\msys\1.0\bin" to PATH environment variable
> of Windows.
> 
> Did you not do the above? Did you do it in a different way?

It sounds like he added them at the end rather than at the beginning.

**Comment by mantri** [comment]
Yes, I did this via powershell terminal. 

$env:PATH += ";C:\Mingw\msys\1.0\bin;C:\MinGW\bin;;C:\Python27"

I had not modified $PATH in the mingw shell before.

PS C:\ovs> $env:PATH
C:\Windows\system32;C:\Windows;C:\Windows\System32\Wbem;C:\Windows\System32\WindowsPowerShell\v1.0\;C:\Program Files (x
86)\Iometer.org\Iometer 2006.07.27;C:\Program Files (x86)\Windows Kits\8.1\Windows Performance Toolkit\;C:\Program Files\Microsoft SQLServer\110\Tools\Binn\;C:\MinGW\bin\;C:\Mingw\msys\1.0\bin\;C:\Python27\

**Comment by mantri** [comment]
@shettyg : I guess that in the $PATH you had pasted above, C:\Windows\System32 is before C:\MinGW\bin. I am not sure how did you get the right sort.exe

**Comment by shettyg** [comment]
Instead of just typing "bash" from MSVC developers command prompt and then doing  a "echo $PATH" inside bash, what happens if you do:
'bash --login' from MSVC developers command prompt and then do "echo $PATH"

**Comment by mantri** [comment]
Yes, with 'bash --login', the $PATH is as we want it to be (mingw/bin the first). 

$ echo $PATH
.:/usr/local/bin:/mingw/bin:/bin:/c/Program Files (x86)/Microsoft Visual Studio
12.0/Common7/IDE/CommonExtensions/Microsoft/TestWindow:/c/Program Files (x86)/Mi
crosoft SDKs/F#/3.1/Framework/v4.0/:/c/Program Files (x86)/MSBuild/12.0/bin:/c/P
rogram Files (x86)/Microsoft Visual Studio 12.0/Common7/IDE/:/c/Program Files (x
86)/Microsoft Visual Studio 12.0/VC/BIN:/c/Program Files (x86)/Microsoft Visual
Studio 12.0/Common7/Tools:/c/Windows/Microsoft.NET/Framework/v4.0.30319:/c/Progr
am Files (x86)/Microsoft Visual Studio 12.0/VC/VCPackages:/c/Program Files (x86)
/HTML Help Workshop:/c/Program Files (x86)/Microsoft Visual Studio 12.0/Team Too
ls/Performance Tools:/c/Program Files (x86)/Windows Kits/8.1/bin/x86:/c/Program
Files (x86)/Microsoft SDKs/Windows/v8.1A/bin/NETFX 4.5.1 Tools/:/c/Windows/syste
m32:/c/Windows:/c/Windows/System32/Wbem:/c/Windows/System32/WindowsPowerShell/v1
.0/:/c/Program Files (x86)/Iometer.org/Iometer 2006.07.27:/c/Program Files (x86)
/Windows Kits/8.1/Windows Performance Toolkit/:/c/Program Files/Microsoft SQL Se
rver/110/Tools/Binn/:/mingw/bin/:/usr/bin/:/c/Python27/:/c/Program Files (x86)/G
it/cmd

C:>bash
bash-3.1$ which sort
/c/Windows/system32/sort.exe
bash-3.1$ exit
exit

C:>bash --login

Administrator@Hemanth-Mantri ~
$ which sort
/bin/sort.exe

**Comment by shettyg** [comment]
I updated the documentation. Closing this. Thanks for the bug report!
